[IVY] fix node injection

vicb opened the pull request.

vicb committed:
docs(ivy): clarify injector API docs

vicb committed:
fix(ivy): walk the node injector tree and then the module injector tree

- `directiveInjector()` is used to inject anything in the directive / component
/ pipe factories so adding `InjectionToken<T>` as a supported token type.
- `getOrCreateInjectable()` should search first in the node injector tree and
then in the module injector tree (was either or before the PR).

mary-poppins commented:
You can preview 99a6554 at https://pr25166-99a6554.ngbuilds.io/.

JoostK commented on packages/core/src/render3/di.ts:
```diff
@@ -433,16 +423,15 @@ function searchMatchesQueuedForCreation<T>(node: LNode, token: any): T|null {
- * @param type The directive type
- * @returns The bloom bit to check for the directive
+ * @param token the injection token
+ * @returns the matching bit to check in the bloom filter of null if the token is not known.
```
`of null` > `or null`

JoostK commented on packages/core/src/render3/di.ts:
```diff
@@ -177,10 +178,12 @@ export function diPublic(def: DirectiveDefInternal<any>): void {
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
+export function directiveInject<T>(
+    token: Type<T>| InjectionToken<T>, flags: InjectFlags.Optional): T|null;
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags): T;
```
This signature is inaccurate (and used to be before this PR), as it depends on the exact flag whether or not the return type should include `null`.

gkalpak replied:
Are there other values (apart from `InjectFlags.Optional`) that can result in `null`?

gkalpak replied:
No worries. Can happen to anyone (even @vicb 😛).

vicb committed:
refactor(ivy): simplify node injector imports

gkalpak commented on packages/core/src/render3/di.ts:
```diff
@@ -174,17 +168,16 @@ export function diPublic(def: DirectiveDefInternal<any>): void {
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
+export function directiveInject<T>(
+    token: Type<T>| InjectionToken<T>, flags: InjectFlags.Optional): T|null;
+export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags): T|null;
```
I think this signature is wrong :grin:
AFAICT, the intention is to have signatures for specific types (that appear above this one) that can produce `null` and have this signature to catch all the rest that cannot produce `null`.

If there are other `InjectFlags` values that can produce `null`, then they should be part of the signature above.
But this signature should not have `null` in the return value.

vicb replied:
I think this line can indeed be removed.
_I don't think there is any way to express what you say using the type system given we use flags_

JoostK replied:
If you remove it fully, it won't accept flags other than `InjectFlags.Optional` at all. [See playground example](https://www.typescriptlang.org/play/index.html#src=export%20const%20enum%20InjectFlags%20%7B%0D%0A%20%20Default%20%3D%200%2C%0D%0A%20%20Optional%20%3D%201%2C%0D%0A%7D%0D%0A%0D%0Aexport%20type%20Type%3CT%3E%20%3D%20Function%3B%0D%0A%0D%0Aexport%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E)%3A%20T%3B%0D%0Aexport%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E%2C%20flags%3A%20InjectFlags.Optional)%3A%20T%7Cnull%3B%0D%0A%2F%2F%20export%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E%2C%20flags%3A%20InjectFlags)%3A%20T%3B%0D%0Aexport%20function%20directiveInject%3CT%3E(token%3A%20Type%3CT%3E%2C%20flags%20%3D%20InjectFlags.Default)%3A%20T%20%7C%20null%20%7B%0D%0A%20%20return%20null%3B%0D%0A%7D%0D%0A%0D%0AdirectiveInject(Array%2C%20InjectFlags.Default)%3B%0D%0A).

vicb replied:
```ts
export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
export function directiveInject<T>(
    token: Type<T>| InjectionToken<T>, flags: InjectFlags = InjectFlags.Default): T|null {
  return getOrCreateInjectable<T>(getOrCreateNodeInjector(), token, flags);
}
```

gkalpak replied:
Yes, the line should not be removed. It needs to be chanfged to return `T` (not `T|null`).
The rest is fine.

mary-poppins commented:
You can preview e705ed6 at https://pr25166-e705ed6.ngbuilds.io/.

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -61,7 +62,7 @@ let nextNgElementId = 0;
  */
 export function bloomAdd(injector: LInjector, type: Type<any>): void {
-  let id: number|undefined = (type as any)[NG_ELEMENT_ID];
+  let id = (type as any)[NG_ELEMENT_ID];
```
Why did you remove `: number|undefined`. Could we add it back?

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -61,7 +62,7 @@ let nextNgElementId = 0;
+  let id = (type as any)[NG_ELEMENT_ID];
 
   // Set a unique ID on the directive type, so if something tries to inject the directive,
   // we can easily retrieve the ID and hash it into the bloom bit that should be checked.
```
since you are already here could we rewrite the code to use bit operations rather than `%` and `<` operations?

mhevery commented:
Could we add tests to `render3/di_spec.ts` for this?

mary-poppins commented:
You can preview d911d61 at https://pr25166-d911d61.ngbuilds.io/.

vicb committed:
fixup! fix(ivy): walk the node injector tree and then the module injector tree

vicb committed:
refactor(ivy): use bit operations in node injector

mary-poppins commented:
You can preview 3ebe4a8 at https://pr25166-3ebe4a8.ngbuilds.io/.

vicb committed:
refactor(ivy): use `element()` where applicable in di_spec

For future ref
Search `elementStart\(([^)]+)\);\s*\n\s*elementEnd\(\);`
Replace `element($1)`

gkalpak commented on packages/core/src/render3/di.ts:
```diff
@@ -174,17 +170,14 @@ export function diPublic(def: DirectiveDefInternal<any>): void {
+ * @returns the value from the injector or `null` when not found
  */
-export function directiveInject<T>(token: Type<T>): T;
-export function directiveInject<T>(token: Type<T>, flags: InjectFlags.Optional): T|null;
```
No, no no. This line should stay :grin:

gkalpak replied:
I.e. this function should be defined as:

```ts
export function directiveInject<T>(token: Type<T>| InjectionToken<T>): T;
export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags.Optional): T|null;
export function directiveInject<T>(token: Type<T>| InjectionToken<T>, flags: InjectFlags): T;
export function directiveInject<T>(
    token: Type<T>| InjectionToken<T>, flags = InjectFlags.Default): T|null {
  return getOrCreateInjectable<T>(getOrCreateNodeInjector(), token, flags);
}
```

vicb replied:
and what if flags = `InjectFlags.Optional | InjectFlags.SkipSelf` ?

mary-poppins commented:
You can preview 5b24146 at https://pr25166-5b24146.ngbuilds.io/.

vicb committed:
build(ivy): fix ci failures

vicb committed:
fixup! refactor(ivy): use bit operations in node injector

mary-poppins commented:
You can preview 5a8d969 at https://pr25166-5a8d969.ngbuilds.io/.

vicb commented:
http://test/OCL:206380401:BASE:206380532:1532729215132:194256de

**merge-assistance**: [golden symbol changes](https://github.com/angular/angular/pull/25166/files#diff-8bec588d7a22fe9ab8f11a8114e80414) are only some renaming - no impact on size.
